File: codes/2024/25-Excellence-BNU-99-21-43 (C++)/workspace/source/backend/controller/LifeTimePoint.cpp
#include "LifeTimePoint.h"

namespace Backend
{
    LifeTimePoint *LifeTimePoint::getUse(LifeTimeIndex *lti)
    {
        return new LifeTimePoint(lti, Type::USE);
    }

    LifeTimePoint *LifeTimePoint::getDef(LifeTimeIndex *lti)
    {
        return new LifeTimePoint(lti, Type::DEF);
    }

    LifeTimePoint::Type LifeTimePoint::getType() const
    {
        return type;
    }

    LifeTimeIndex *LifeTimePoint::getIndex() const
    {
        return ltindex;
    }

    bool LifeTimePoint::isDef() const
    {
        return type == Type::DEF;
    }

    bool LifeTimePoint::isUse() const
    {
        return type == Type::USE;
    }

    bool LifeTimePoint::isTruePoint() const
    {
        AsmInstBasic *inst = getIndex()->getSrcInst();
        if ((inst->getOpcodeBasic() == AsmInstBasic::OpcodeBasic::AsmInstJump) || (inst->getOpcodeBasic() == AsmInstBasic::OpcodeBasic::AsmInstLabel))
            return false;
        return true;
    }

    std::string LifeTimePoint::toString() const
    {
        return getIndex()->toString() + ":" + (type == Type::DEF ? "DEF" : "USE");
    }

    bool LifeTimePoint::operator<(const LifeTimePoint &other) const
    {
        return *(this->ltindex) < *(other.getIndex());
    }

    bool LifeTimePoint::operator==(const LifeTimePoint &other) const
    {
        return *(this->ltindex) == *(other.getIndex());
    }
}

File: codes/2023/2-First-BIT (Java)/src/main/java/cn/edu/bit/newnewcc/backend/asm/controller/LifeTimePoint.java
package cn.edu.bit.newnewcc.backend.asm.controller;

import cn.edu.bit.newnewcc.backend.asm.instruction.AsmJump;
import cn.edu.bit.newnewcc.backend.asm.instruction.AsmLabel;

public class LifeTimePoint implements Comparable<LifeTimePoint> {
    public enum Type {
        USE, DEF
    }

    private final LifeTimeIndex index;
    private final Type type;

    private LifeTimePoint(LifeTimeIndex index, Type type) {
        this.index = index;
        this.type = type;
    }
    public static LifeTimePoint getUse(LifeTimeIndex index) {
        return new LifeTimePoint(index, Type.USE);
    }
    public static LifeTimePoint getDef(LifeTimeIndex index) {
        return new LifeTimePoint(index, Type.DEF);
    }

    public Type getType() {
        return type;
    }

    public LifeTimeIndex getIndex() {
        return index;
    }

    public boolean isDef() {
        return type == Type.DEF;
    }

    public boolean isUse() {
        return type == Type.USE;
    }

    public boolean isTruePoint() {
        var inst = getIndex().getSourceInst();
        return !(inst instanceof AsmLabel || inst instanceof AsmJump);
    }

    @Override
    public String toString() {
        return getIndex() + ":" + getType();
    }

    @Override
    public int compareTo(LifeTimePoint o) {
        return index.compareTo(o.getIndex());
    }

}
